---
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval = FALSE)
```

# Results 

```{r}

library(knitr)

## Create table for Mapping Rate
mapping_table <- data.frame(
  Condition = c("treated1", "treated3", "control2", "control1", "treated2", "control3"),
  Mapping_Rate = c("90.854%", "92.0915%", "90.6273%", "89.3778%", "91.3572%", "91.3243%"),
  stringsAsFactors = FALSE
)


## Print the resulting table
print(mapping_table)

   Condition Mapping_Rate
1  treated1      90.854%
2  treated3     92.0915%
3  treated2     91.3572%
4  control2     90.6273%
5  control1     89.3778%
6  control3     91.3243%

## Loading Required Packages
install.packages("tximport")
install.packages("DESeq2")
library(tximport)
library(DESeq2)

## Setting Sample Information
sample_names <- c("treated1", "treated2", "treated3", "control1", "control2", "control3")
sample_condition <- c(rep('Treated',3),rep('Control',3))
folder_path <- "/Users/mariagarzon/Downloads/final_ngs/"
sample_paths <- paste0(folder_path, sample_names, "/quant.sf")

files <- file.path(sample_paths)
names(files) <- sample_names

tx2gene <- read.table("/Users/mariagarzon/Downloads/final_ngs/tx2gene.csv",header=F,sep=",") ## CSV file containing transcript-to-gene mapping information.

## Importing Transcript Quantification Data
txi <- tximport(files, type="salmon", tx2gene=tx2gene)

## Creating Sample Metadata
samples <- data.frame(sample_names=sample_names,condition=sample_condition)
row.names(samples) <- sample_names

## Create DESeqDataSet object
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~ condition)
class(ddsTxi)

keep <- rowSums(counts(ddsTxi)) >= 10 # filter out genes with low counts
dds <- ddsTxi[keep,]

## Differential Expression Analysis
ddsTxi <- DESeq(ddsTxi)
class(ddsTxi)

## Extracting results from the DESeq2 analysis using the “results” function.
## The results function is applied to ddsTxi to extract the differential expression results, 
## specifically contrasting the "Treated" and "Control" conditions.
res <- results(ddsTxi, contrast = c('condition','Treated','Control') )
class(res) 

## Soring the data in ascending p-value
resOrdered <- res[order(res$pvalue),] 
head(resOrdered,10) # View the 10 most significant genes

## Perform log-fold change shrinkage for more stable estimates of differential expression
res.shrunk <- lfcShrink(ddsTxi, coef=2, type='apeglm')
res.shrunkOrdered <- res.shrunk[order(res.shrunk$pvalue),]

resOrdered
res.shrunkOrdered # ordered by adjusted p-value


## MA plots for raw and shrunk results 
plotMA(res)
plotMA(res.shrunk)

### The shrunken MA plot represents the differential expression analysis with log-fold change shrinkage, which helps to stabilize the estimates of differential expression.
### The shrunken MA plot has a tighter distribution of points around the center line compared to the raw MA plot.

## Top Differentially Expressed Genes based on adjusted p-value
top10genes <- head(res.shrunkOrdered, 10)
#### Top 10 differentially expressed genes according to adjusted p-value 
top10genes
log2 fold change (MAP): condition Treated vs Control 
Wald test p-value: condition Treated vs Control 
DataFrame with 10 rows and 5 columns
baseMean log2FoldChange     lfcSE       pvalue         padj
<numeric>      <numeric> <numeric>    <numeric>    <numeric>
ENSG00000196396.10   6561.59       1.154994 0.0437049 3.23202e-155  4.30279e-151
ENSG00000175334.8    6404.59       1.658194 0.0646399 2.12129e-146  1.41204e-142
ENSG00000206286.11   2775.24      -1.384980 0.0596141 1.23519e-120  5.48135e-117
ENSG00000128595.17  19576.72       1.478563 0.0710831  2.26447e-97  7.53672e-94
ENSG00000101384.12  11621.67       1.286671 0.0628566  1.21088e-94  3.22408e-91
ENSG00000124333.16   2706.82       1.458485 0.0736103  1.36364e-88  3.02568e-85
ENSG00000117632.23  16592.10       1.329956 0.0717960  7.99474e-78  1.52049e-74
ENSG00000143384.13  21042.16       1.024778 0.0562517  2.59931e-75  4.32558e-72
ENSG00000145919.11   2678.78       1.265271 0.0721386  4.17755e-70  6.17952e-67
ENSG00000144028.15  12238.69      -0.812422 0.0466342  4.23595e-69  5.63932e-66

### ENSG00000196396.10, ENSG00000175334.8,ENSG00000128595.17,ENSG00000101384.12, and ENSG00000124333.16,ENSG00000117632.23, all have a large log fold-change and an extremely small adjusted p-value, indicating a significant and substantial difference in expression between the treated and control conditions. ENSG00000206286.11  has a moderate negative log fold-change and a very small adjusted p-value. Although it has a significant statistical difference between the treated and control conditions, further examination is required. ENSG00000143384.13 and ENSG00000145919.11 have a moderate negative log fold-change and a very small adjusted p-value. Although it has a significant statistical difference between the treated and control conditions, further examination is required to determine its biological relevance.

## Number of statistically significant genes at an FDR of 0.05
chosen_FDR <- 0.05
significant_genes <- sum(res.shrunkOrdered$padj < chosen_FDR, na.rm = TRUE)
significant_genes = 2815

## Number of biologically relevant differentially expressed genes based ontwo-fold change in gene expression
fold_change_threshold <- 2
relevant_genes <- sum(abs(res.shrunkOrdered$log2FoldChange) >= log2(fold_change_threshold), na.rm = TRUE)
relevant_genes = 159

## Estimate dispersions using mean method
ddsTxi_mean <- estimateDispersions(ddsTxi, fitType="mean")
plotDispEsts(ddsTxi_mean, main = "Mean Method")

### The dispersion estimate plot shows the gene-wise estimates(black),the fitted values(red),and the final maximuma posteriori estimates used in testing(blue). The red line in the dispersion plot is straight horizontally and doesn't curve, it indicates that the mean-dispersion relationship is not  varying across the range of mean expression values. In other words, the dispersion estimates are relatively stable and consistent across different expression levels.

## Visualization of Histogram
library(ggplot2)
ggplot(as.data.frame(res.shrunk),aes(pvalue)) + geom_histogram(fill="light blue",color='black')
#### The pattern observed in the histogram has an enrichment of low p-values.  This is the expected result if there is a large class of differentially  expressed genes between treatment and control. This suggests that knockout of the genn does affect gene expression as shown with the low p-values (rejecting the null hypothesis that sugar composition does not affect gene expression). This shows that certain genes are differentially expressed between control and treatment

## Write the DESeq2 results for all genes to a tab-delimited file
## Specify the file path where you want to save the results
output_file <- "/Users/mariagarzon/Downloads/final_ngs/DESeq2_results.txt"
write.table(as.data.frame(res.shrunkOrdered), file = output_file, sep = "\t", quote = FALSE, row.names = FALSE)
```